File: backend/src/main/java/ca/metricalsky/winston/entity/CommentEntity.java
package ca.metricalsky.winston.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import lombok.Getter;
import lombok.Setter;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.OffsetDateTime;
import java.util.List;

import static jakarta.persistence.CascadeType.MERGE;
import static jakarta.persistence.CascadeType.PERSIST;

@Entity
@Table(name = "comments")
@Getter
@Setter
public class CommentEntity {

    @Id
    @Column(name = "id")
    private String id;

    @Column(name = "video_id")
    private String videoId;

    @Column(name = "parent_id")
    private String parentId;

    @ManyToOne
    @JoinColumn(name = "author_id", referencedColumnName = "id")
    private AuthorEntity author;

    @Column(name = "text_display")
    private String textDisplay;

    @Column(name = "text_original")
    private String textOriginal;

    @Column(name = "total_reply_count")
    private Long totalReplyCount;

    @Column(name = "published_at")
    private OffsetDateTime publishedAt;

    @Column(name = "updated_at")
    private OffsetDateTime updatedAt;

    @UpdateTimestamp
    @Column(name = "last_fetched_at")
    private OffsetDateTime lastFetchedAt;

    @OneToMany(cascade = {PERSIST, MERGE})
    @JoinColumn(name = "parent_id", referencedColumnName = "id")
    private List<CommentEntity> replies;

}

File: database/src/main/kotlin/ca/metricalsky/winston/database/entity/comment/CommentEntity.kt
package ca.metricalsky.winston.database.entity.comment

import ca.metricalsky.winston.database.entity.author.AuthorEntity
import jakarta.persistence.*
import org.hibernate.annotations.UpdateTimestamp
import java.time.OffsetDateTime

@Entity
@Table(name = "comments")
class CommentEntity {

    @Id
    @Column(name = "id")
    var id: String? = null

    @Column(name = "video_id")
    var videoId: String? = null

    @Column(name = "parent_id")
    var parentId: String? = null

    @ManyToOne
    @JoinColumn(name = "author_id", referencedColumnName = "id")
    var author: AuthorEntity? = null

    @Column(name = "text_display")
    var textDisplay: String? = null

    @Column(name = "text_original")
    var textOriginal: String? = null

    @Column(name = "like_count")
    var likeCount: Long? = null

    @Column(name = "total_reply_count")
    var totalReplyCount: Long? = null

    @Column(name = "published_at")
    var publishedAt: OffsetDateTime? = null

    @Column(name = "updated_at")
    var updatedAt: OffsetDateTime? = null

    @UpdateTimestamp
    @Column(name = "last_fetched_at")
    var lastFetchedAt: OffsetDateTime? = null

    @OneToOne
    @JoinColumn(name = "id", referencedColumnName = "comment_id")
    var properties: CommentPropertiesEntity? = null

    @OneToMany(cascade = [CascadeType.PERSIST, CascadeType.MERGE])
    @JoinColumn(name = "parent_id", referencedColumnName = "id")
    var replies: MutableList<CommentEntity?>? = null
}
